cloud/aws/actuators/machine/actuator.go

```diff
@@ -1,63 +1,504 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var securityGroupIds []*string
+	for _, g := range describeSecurityGroupsOutput.SecurityGroups {
+		groupID := *g.GroupId
+		securityGroupIds = append(securityGroupIds, &groupID)
+	}
+
+	// build list of networkInterfaces (just 1 for now)
+	var networkInterfaces = []*ec2.InstanceNetworkInterfaceSpecification{
+		{
+			DeviceIndex:              aws.Int64(machineProviderConfig.DeviceIndex),
+			AssociatePublicIpAddress: aws.Bool(true),
+			SubnetId:                 describeSubnetsResult.Subnets[0].SubnetId,
+			Groups:                   securityGroupIds,
+		},
+	}
+
+	// Add tags to the created machine
+	tagList := []*ec2.Tag{}
+	for _, tag := range machineProviderConfig.Tags {
+		tagList = append(tagList, &ec2.Tag{Key: aws.String(tag.Name), Value: aws.String(tag.Value)})
+	}
+	tagList = append(tagList, []*ec2.Tag{
+		{Key: aws.String("clusterid"), Value: aws.String(cluster.Name)},
+		{Key: aws.String("kubernetes.io/cluster/" + cluster.Name), Value: aws.String(cluster.Name)},
+		{Key: aws.String("Name"), Value: aws.String(machine.Name)},
+	}...)
+
+	tagInstance := &ec2.TagSpecification{
+		ResourceType: aws.String("instance"),
+		Tags:         tagList,
+	}
+	tagVolume := &ec2.TagSpecification{
+		ResourceType: aws.String("volume"),
+		Tags:         tagList[0:1],
+	}
+
+	userData := []byte{}
+	if machineProviderConfig.UserDataSecret != nil {
+		userDataSecret, err := a.kubeClient.CoreV1().Secrets(machine.Namespace).Get(machineProviderConfig.UserDataSecret.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if data, exists := userDataSecret.Data[userDataSecretKey]; exists {
+			userData = data
+		} else {
+			glog.Warning("Secret %v/%v does not have %q field set. Thus, no user data applied when creating an instance.", machine.Namespace, machineProviderConfig.UserDataSecret.Name, userDataSecretKey)
+		}
+	}
+
+	userDataEnc := base64.StdEncoding.EncodeToString(userData)
+
+	inputConfig := ec2.RunInstancesInput{
+		ImageId:      describeAMIResult.Images[0].ImageId,
+		InstanceType: aws.String(machineProviderConfig.InstanceType),
+		// Only a single instance of the AWS instance allowed
+		MinCount: aws.Int64(1),
+		MaxCount: aws.Int64(1),
+		KeyName:  machineProviderConfig.KeyName,
+		IamInstanceProfile: &ec2.IamInstanceProfileSpecification{
+			Name: aws.String(*machineProviderConfig.IAMInstanceProfile.ID),
+		},
+		TagSpecifications: []*ec2.TagSpecification{tagInstance, tagVolume},
+		NetworkInterfaces: networkInterfaces,
+		UserData:          &userDataEnc,
+	}
+
+	runResult, err := client.RunInstances(&inputConfig)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create EC2 instance: %v", err)
+	}
+
+	if runResult == nil || len(runResult.Instances) != 1 {
+		mLog.Errorf("unexpected reservation creating instances: %v", runResult)
+		return nil, fmt.Errorf("unexpected reservation creating instance")
+	}
+	return runResult.Instances[0], nil
```

Comment:
One issue we have right now with how this works, is that usually at the end of the RunInstances call, the instance state is not yet 'Running'. Rather it is 'Pending'. Therefore the status is not yet complete (like PublicIP). One possible fix would be to wait and poll here until the instance is in the Running state. The other approach is to return an error so that the machine is requeued and in the exists function return false if the status doesn't match. That would require making the create re-entrant so we can just update the status.

Reply:
Put TODO comment in the code.